l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,444 @@
+            weight: 0,
+            pubKey: Secp256k1PublicKey({tag: bytes1(0), x: bytes32(0)})
+        }),
+            attesterLastUpdateCommit: attestersCommit,
+            validatorLatest: ValidatorAttr({
+            active: true,
+            pendingRemoval: false,
+            weight: _validatorWeight,
+            pubKey: _validatorPubKey,
+            pop: _validatorPoP
+        }),
+            validatorSnapshot: ValidatorAttr({
+            active: false,
+            pendingRemoval: false,
+            weight: 0,
+            pubKey: BLS12_381PublicKey({a: bytes32(0), b: bytes32(0), c: bytes32(0)}),
+            pop: BLS12_381Signature({a: bytes32(0), b: bytes16(0)})
+        }),
+            validatorLastUpdateCommit: validatorsCommit
+        });
+
+        emit NodeAdded(
+            _nodeOwner,
+            _validatorWeight,
+            _validatorPubKey,
+            _validatorPoP,
+            _attesterWeight,
+            _attesterPubKey
+        );
+    }
+
+    /// @notice Deactivates a node, preventing it from participating in committees.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be inactivated.
+    function deactivate(address _nodeOwner) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
+        if (_finalizeNodeRemoval(_nodeOwner, node)) {
+            return;
+        }
+
+        _ensureAttesterSnapshot(node);
+        node.attesterLatest.active = false;
+        _ensureValidatorSnapshot(node);
+        node.validatorLatest.active = false;
+
+        emit NodeDeactivated(_nodeOwner);
+    }
+
+    /// @notice Activates a previously inactive node, allowing it to participate in committees.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be activated.
+    function activate(address _nodeOwner) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
+        if (_finalizeNodeRemoval(_nodeOwner, node)) {
+            return;
+        }
+
+        _ensureAttesterSnapshot(node);
+        node.attesterLatest.active = true;
+        _ensureValidatorSnapshot(node);
+        node.validatorLatest.active = true;
+
+        emit NodeActivated(_nodeOwner);
+    }
+
+    /// @notice Removes a node from the registry.
+    /// @dev Only callable by the contract owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be removed.
+    function remove(address _nodeOwner) external onlyOwner {
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
```

Comment:
Shouldn't `_finalizeNodeRemoval` be called here as well? 

I'm just thinking how intuitive this API is. I understand you can't immediately delete records because because they might still be part of the committee, so if I want to know who are the active members I need to look at the `commit`, the `lastUpdateCommit`, the `active` and the `pendingRemoval` of either the `lastest` or the `snapshot`. I also understand that `commitValidators` and `commitAttesters` can happen independently and we only want to remove when both indicate the `pendingRemoval` is committed. What I wouldn't understand as a user is why I should call `change<something>` to get the record removed, why couldn't I call `remove` again to get rid of it, which has the most intuitive name for this purpose? That is, if I call `remove` and it's already pending removal in the snapshot, then we can just delete it.

Reply:
As you referred to, `_finalizeNodeRemoval` wasn’t called here because `remove` was intended for the initial phase of the removal.

I agree that overloading this function rather than solely relying on other calls is probably better. However, this raises the question of whether to continue using the other calls to finalize removals. I think that not doing so is problematic because it suggests that changes can still be made to a node that should be deleted.

P.S. to reduce confusion, I've termed the initial phase as "remove" and the finalization as "delete".

Changes here: https://github.com/matter-labs/era-contracts/pull/555/commits/a13bcf6a0a9b16901f43af1071d3d18e488616df